### PR TITLE
Customized L1T NanoAOD

### PR DESCRIPTION
**Pull request by lathomas**
#### PR description:

This package allows to save Level 1 Trigger information (unpacked or reemulated) in a standard NANOAOD format. A few examples of cmsDriver commands are provided for Run 3 data and simulated samples in the README file. 

The current PR focuses on Phase 1 L1 objects (unpacked or reemulated), and calorimeter trigger primitives and trigger towers. It is expected to be completed in the future with additional information (e.g. on muon TPs). 

#### PR validation:

The `cmsDriver.py` commands in the README are found to properly run.

### CODE REVIEW

**Comment by vlimant** [comment]
assign xpog
indeed

**Comment by vlimant** [comment]
The first two recommendations are to follow the `NANO:@L1T` and to add a test workflow in the nano matrix. Can you please explain the difference with respect to #40663 ?

**Comment by lathomas** [comment]
https://github.com/cms-sw/cmssw/pull/40663 allows to add the unpacked L1 objects to central NANO, starting from MINIAOD. This PR is much more generic: it allows to define a custom NANO format, that stores both emulated and unpacked L1 objects, but also the Trigger Primitives (calo only for now). This package aims at replacing https://github.com/cms-sw/cmssw/tree/master/L1Trigger/L1TNtuples/

**Comment by vlimant** [comment]
please see #43065 for the test and the NANO:@L1T syntax (for lack of better set of instructions)

**Comment by artlbv** [comment]
@lathomas I used this as a reference:

 https://github.com/cms-sw/cmssw/pull/42238

And in the PR from above it must be here:

<https://github.com/cms-sw/cmssw/pull/43065/files#diff-2380e1dfdfadf32cba9abec31248637f27f5ce4dc6fc0b2179de7232c78728ba>
[3865030.jpeg]
Fixes/impovements towards the first production version of the custom muon DPG NANO by battibass · Pull Request #43065 · cms-sw/cmssw<https://github.com/cms-sw/cmssw/pull/43065/files#diff-2380e1dfdfadf32cba9abec31248637f27f5ce4dc6fc0b2179de7232c78728ba>
github.com<https://github.com/cms-sw/cmssw/pull/43065/files#diff-2380e1dfdfadf32cba9abec31248637f27f5ce4dc6fc0b2179de7232c78728ba>

File: DPGAnalysis/L1TNanoAOD/python/l1tnanotables_phase2_cff.py
Comment (artlbv) [inline DPGAnalysis/L1TNanoAOD/python/l1tnanotables_phase2_cff.py:1]: is this supposed to be included in this PR? :)

File: Configuration/PyReleaseValidation/python/relval_nano.py [hunk 1]
```diff
@@ -167,6 +167,13 @@ def subnext(self):
                                    '--datatier':'NANOAOD',
                                    '--eventcontent':'NANOAOD'}])
 
+steps['l1DPGNANO_data13.0']=merge([{'-s' : 'RAW2DIGI,NANO:@L1DPG',
```
Comment (vlimant) [inline Configuration/PyReleaseValidation/python/relval_nano.py:170]: can you please do something like `muPOGNANO_data13.0` (merge with `steps['NANO_data13.0']`) instead ? I could ask that you modify `muDPGNANO_data13.0` and `muDPGNANOBkg_data13.0` the same way to have it consistent, but this is not necessary here

File: DPGAnalysis/L1TNanoAOD/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+ ### Run PAT, NANO and save all calo information (TPs, TT, CaloCluster) and L1T objects, both emulated and unpacked (data) 
+ 
+
+    cmsDriver.py customL1toNANO --conditions auto:run3_data_prompt -s RAW2DIGI,L1Reco,PAT,NANO,USER:DPGAnalysis/L1TNanoAOD/l1tNano_cff.l1tNanoTask --datatier NANOAOD --eventcontent NANOAOD --data --process customl1nano --scenario pp --era Run3 --customise Configuration/DataProcessing/RecoTLR.customisePostEra_Run3,PhysicsTools/NanoAOD/nano_cff.nanoL1TrigObjCustomizeFull,DPGAnalysis/L1TNanoAOD/l1tNano_cff.addCaloFull,L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAWsimHcalTP -n 100 --filein /store/data/Run2022D/EGamma/RAW-RECO/ZElectron-27Jun2023-v2/2810003/06757985-055e-4c64-bbe3-187858ea2abf.root --fileout file:out.root --python_filename=customl1nano.py
```
Comment (vlimant) [inline DPGAnalysis/L1TNanoAOD/README.md:44]: change to use the `NANO:@L1DPG` instead, and everywhere above where relevant

File: DPGAnalysis/L1TNanoAOD/plugins/CaloTPTableProducer.cc [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+CaloTPTableProducer::endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
+{
+}
+*/
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void CaloTPTableProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
```
Comment (vlimant) [inline DPGAnalysis/L1TNanoAOD/plugins/CaloTPTableProducer.cc:252]: please add the proper parameter validation : see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

**Comment by aloeliger** [comment]
@lathomas Could you real quickly hit the code-formatting (`scram build code-format`) on this?

**Comment by vlimant** [comment]
enable nano

**Comment by vlimant** [comment]
please test

File: DPGAnalysis/L1TNanoAOD/plugins/CaloTPTableProducer.cc
Comment (aloeliger) [inline DPGAnalysis/L1TNanoAOD/plugins/CaloTPTableProducer.cc:1]: I think this is fine from me, could you just get rid of some of the commented or unused functions, and set the destructor to default here?

**Comment by vlimant** [comment]
please go ahead with a shared signature, if L1T people agree too.

**Comment by aloeliger** [comment]
please test
- re-triggering tests after cosmetic changes. I think L1T is fine with this and can sign when done.

**Comment by aloeliger** [comment]
@lathomas Could you rebase this PR on top of the current state of CMSSW?

**Comment by aloeliger** [comment]
please test

**Comment by AdrianoDee** [comment]
+pdmv
(wf wise everything is fine)

**Comment by AdrianoDee** [comment]
> The following packages do not have a category, yet:
> 
> DPGAnalysis/L1TNanoAOD Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@lathomas since you are adding a new sub-package here would you mind taking care of adding it to the bot categories? I imagine this would need to be assigned to both `xpog` and `l1`.

**Comment by AdrianoDee** [comment]
Sorry for the noise, I see @aloeliger already took care of it.

File: Configuration/PyReleaseValidation/python/relval_nano.py [hunk 4]
```diff
@@ -152,24 +152,28 @@ def subnext(self):
...
+                                       '--era' : 'Run3',
+                                       '--datatier':'NANOAOD',
+                                       '--eventcontent':'NANOAOD'}])
 
 steps['muPOGNANO_data13.0']=merge([{'-s' : 'NANO:@MUPOG,DQM:@nanoAODDQM', '-n' : '1000'},
-                                    steps['NANO_data13.0']])
+                                   steps['NANO_data13.0']])
 
+steps['l1DPGNANO_data13.0']=merge([{'-s' : 'RAW2DIGI,NANO:@L1DPG',
+                                    '--conditions':'auto:run3_data',
```
Comment (vlimant) [inline Configuration/PyReleaseValidation/python/relval_nano.py:173]: the `--conditions` can be removed, as it is already in `steps['NANO_data13.0']`

**Comment by vlimant** [comment]
please test

**Comment by aloeliger** [comment]
+l1
- re-sign

**Comment by vlimant** [comment]
+1

**Comment by srimanob** [comment]
@lathomas 
It seems this PR needs rebase.

**Comment by lathomas** [comment]
@vlimant @srimanob Should I rebase now or do you anticipate more changes conflicting with this in the near future? 

Thanks. 

Laurent

**Comment by vlimant** [comment]
the conflicts must be in autoNano and reval_nano, and should be trivial to fix. more will come, the longer it takes to finalise this PR

**Comment by lathomas** [comment]
@AdrianoDee @srimanob can you take a look and sign if happy? Thanks !

**Comment by AdrianoDee** [comment]
Just a remark for the future: unless really needed I would append the new wfs rather than placing them in the middle resulting in a change to the wf numbering. Maybe in this case it is just an artifact due to the multiple concurrent PRs and the need to rebase, but still.

**Comment by AdrianoDee** [comment]
+pdmv

**Comment by antoniovilela** [comment]
+1

**Comment by vlimant** [comment]
I am puzzled how tests succeeded here and lead to #44666 ...

**Pull request merged**